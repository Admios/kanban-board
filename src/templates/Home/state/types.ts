export interface Todo {
  id: string;
  columnId: string;
  text: string;
  position: number;
  owner: string;
}

export interface Column {
  id: string;
  name: string;
  position: number;
  color: string;
  owner: string;
}
