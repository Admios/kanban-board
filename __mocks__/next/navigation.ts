const push = jest.fn();
export const useRouter = () => ({ push });
export const redirect = jest.fn();
export const navigation = { push, redirect };
