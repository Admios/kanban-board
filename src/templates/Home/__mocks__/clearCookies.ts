export const clearCookies = jest.fn().mockResolvedValue(undefined);
