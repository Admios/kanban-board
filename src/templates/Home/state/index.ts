import { ColumnDTO as DbColumn } from "@/model/Column/ColumnDTO";
import { TaskDTO as DBTodo } from "@/model/Task/TaskDTO";
import { UserDTO } from "@/model/User/UserDTO";
import { Immutable, produce } from "immer";
import { v4 as uuid } from "uuid";
import { StateCreator, create } from "zustand";
import { Column, Todo } from "./types";

export type { Column, Todo } from "./types";

type HomeState = Immutable<{
  todos: Record<string, Todo[]>;
  columns: Record<string, Column>;
  user?: UserDTO;
}>;

interface HomeActions {
  initialize(
    initialTodos: DBTodo[],
    initialColumns: DbColumn[],
    initialUser?: UserDTO,
  ): void;
  addTodo(newTodo: Omit<Todo, "id">): Todo;
  moveTodo(
    newTodo: Todo,
    fromColumnId: string,
    toColumnId: string,
    position: number,
  ): void;
  addColumn(newColumn: Omit<Column, "id" | "position">): Column;
}

const stateCreator: StateCreator<HomeState & HomeActions> = (set, get) => ({
  todos: {},
  columns: {},

  initialize(initialTodos, initialColumns, initialUser) {
    const columnMap: Record<string, Column> = {};
    const todosMap: Record<string, Todo[]> = {};
    const backendIdToIdMap = new Map<string, string>();

    initialColumns.forEach((backendColumn) => {
      const result: Column = {
        ...backendColumn,
        id: uuid(),
        backendId: backendColumn.id,
      };
      columnMap[result.id] = result;
      if (result.backendId) {
        backendIdToIdMap.set(result.backendId, result.id);
      }
    });

    initialTodos.forEach((backendTodo) => {
      const columnId =
        backendIdToIdMap.get(backendTodo.columnId) ?? backendTodo.columnId;
      if (!todosMap[columnId]) {
        todosMap[columnId] = [];
      }
      todosMap[columnId].push({
        ...backendTodo,
        columnId,
        id: uuid(),
        backendId: backendTodo.id,
      });
    });

    set({ todos: todosMap, columns: columnMap, user: initialUser });
  },

  addTodo: (newTodo) => {
    const result = { ...newTodo, id: uuid() };
    const todos = produce(get().todos, (draft) => {
      if (!draft[newTodo.columnId]) {
        draft[newTodo.columnId] = [];
      }

      draft[newTodo.columnId].push(result);
    });

    set({ todos });
    return result;
  },

  moveTodo: (newTodo, fromColumnId, toColumnId, position) => {
    const todos = produce(get().todos, (draftState) => {
      const sourceColumn = (draftState[fromColumnId] ?? []).filter(
        (todo) => todo.id !== newTodo.id,
      );

      let destinationColumn = draftState[toColumnId] ?? [];
      if (fromColumnId === toColumnId) {
        destinationColumn = sourceColumn;
      }

      if (position <= destinationColumn.length) {
        destinationColumn.splice(position, 0, newTodo);
      } else {
        destinationColumn.push(newTodo);
      }

      draftState[fromColumnId] = sourceColumn;
      draftState[toColumnId] = destinationColumn;
    });

    set({ todos });
  },

  addColumn: (newColumn) => {
    const currentState = get();
    const newId = uuid();
    const todos = produce(currentState.todos, (draft) => {
      draft[newId] = [];
    });
    const columns = produce(currentState.columns, (draft) => {
      const position = Object.keys(currentState.columns).length;
      draft[newId] = { ...newColumn, id: newId, position };
    });

    set({ columns, todos });
    return columns[newId];
  },
});

export const useZustand = create(stateCreator);
