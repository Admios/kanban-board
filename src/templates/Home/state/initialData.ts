import { Column, Todo } from "./types";
import { v4 as uuid } from "uuid";

export const columnNew: Column = {
  name: "New",
  position: 0,
  color: "black",
  id: "columnNew",
};

export const columnNewTodos: Todo[] = [
  {
    id: uuid(),
    text: "Task 17",
  },
  {
    id: uuid(),
    text: "Task 28",
  },
  {
    id: uuid(),
    text: "Task 39",
  },
];

export const columnInProgress: Column = {
  name: "In Progress",
  position: 1,
  color: "red",
  id: "columnInProgress",
};

export const columnInProgressTodos: Todo[] = [
  {
    id: uuid(),
    text: "Task 45",
  },
  {
    id: uuid(),
    text: "Task 56",
  },
];

export const columnInReview: Column = {
  name: "Review",
  position: 2,
  color: "green",
  id: "columnReview",
};

export const columnInReviewTodos: Todo[] = [
  {
    id: uuid(),
    text: "Task 87",
  },
  {
    id: uuid(),
    text: "Task 29",
  },
  {
    id: uuid(),
    text: "Task 63",
  },
  {
    id: uuid(),
    text: "Task 4",
  },
];

export const columnDone: Column = {
  name: "Done",
  position: 3,
  color: "blue",
  id: "columnDone",
};

export const columnDoneTodos: Todo[] = [
  {
    id: uuid(),
    text: "Task 7",
  },
  {
    id: uuid(),
    text: "Task 8",
  },
  {
    id: uuid(),
    text: "Task 9",
  },
];
