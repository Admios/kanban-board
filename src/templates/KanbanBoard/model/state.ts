import { StateDTO } from "@/model/State";
import { TaskDTO } from "@/model/Task";
import { UserDTO } from "@/model/User";
import { Immutable } from "immer";
import { State, Task } from "./types";

export type KanbanState = Immutable<{
  /**
   * The order of the states in the board.
   */
  statesOrder: string[];

  /**
   * The order of the tasks in each state.
   */
  tasksOrder: Record<string, string[]>;

  /**
   * Key: taskId. Value: Task.
   */
  tasks: Record<string, Task>;

  /**
   * Key: stateId. Value: State.
   */
  states: Record<string, State>;

  /**
   * The user that is currently logged in
   */
  user?: UserDTO;
}>;

export interface KanbanActions {
  setUser(user?: UserDTO): void;
  initialize(initialTasks: TaskDTO[], initialStates: StateDTO[]): void;
  addTask(newTask: Task): Task;
  moveTask(
    newTask: Task,
    fromStateId: string,
    toStateId: string,
    position: number,
  ): void;
  addState(newState: State): State;
  editTask(id: string, updatedValues: Partial<Task>): void;
  deleteTask: (id: string) => void;

  clear(): void;
}
