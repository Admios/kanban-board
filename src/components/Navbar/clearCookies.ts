"use server";

import { cookies } from "next/headers";

export async function clearCookies() {
  cookies()
    .getAll()
    .forEach((cookie) => {
      cookies().delete(cookie.name);
    });
}
