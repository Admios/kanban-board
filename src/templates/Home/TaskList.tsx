import { Box, Flex, Heading, useDisclosure } from "@chakra-ui/react";
import { useMemo, useState } from "react";
import { AddColumnModal } from "./AddColumnModal";
import { Column } from "./Column";
import { Header } from "./Header";
import { Column as ColumnType, useZustand } from "./state";
import { AddTodoModal } from "./AddTodoModal";

export function TaskList() {
  const addColumn = useZustand((store) => store.addColumn);
  const addTodo = useZustand((store) => store.addTodo);
  const columns = useZustand((store) => store.columns);
  const [addTodoModalColId, setTodoModalColId] = useState<string | undefined>();
  const {
    isOpen: isColumnDialogOpen,
    onOpen: onOpenColumnDialog,
    onClose: onCloseColumnDialog,
  } = useDisclosure();

  const sortedColumns = useMemo(() => {
    return Object.values(columns).sort(
      (valueA, valueB) => valueA.position - valueB.position,
    );
  }, [columns]);

  const handleCreateRandomTasks = () => {
    const randomTasks = new Set<string>();
    let firstColumn: ColumnType;

    if (!sortedColumns.length) {
      firstColumn = addColumn({
        name: "Random Column",
        backendId: null,
        color: "black",
      });
    } else {
      firstColumn = sortedColumns[0];
    }

    while (randomTasks.size < 10) {
      randomTasks.add(`Random Task ${Math.floor(Math.random() * 100)}`);
    }

    randomTasks.forEach((task) =>
      addTodo({ text: task, columnId: firstColumn.id, backendId: null }),
    );
  };

  return (
    <Box margin="4">
      <Header
        handleCreateRandomTasks={handleCreateRandomTasks}
        onOpenColumnDialog={onOpenColumnDialog}
      />
      <Heading mx="auto" paddingBottom="2">
        Board
      </Heading>

      <Flex direction={"row"} gap="2">
        {sortedColumns.map((value) => (
          <Column
            key={value.name}
            colId={value.id}
            colTitle={value.name}
            color={value.color}
            onOpenCreateTodoModal={() => setTodoModalColId(value.id)}
          />
        ))}
      </Flex>

      <AddColumnModal
        isOpen={isColumnDialogOpen}
        onClose={onCloseColumnDialog}
      />
      <AddTodoModal
        isOpen={!!addTodoModalColId}
        onClose={() => setTodoModalColId(undefined)}
        columnId={addTodoModalColId}
      />
    </Box>
  );
}
