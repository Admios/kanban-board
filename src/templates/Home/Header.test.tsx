import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { useRouter } from "next/navigation";
import { TaskList } from "./TaskList";
import { clearCookies } from "./clearCookies";
import { useZustand } from "./state";

jest.mock("./serverActions.ts");
jest.mock("./clearCookies.ts");

it("should launch login when button is pressed", async () => {
  const { push } = (useRouter as jest.Mock)();
  const { container } = render(<TaskList />);
  expect(container).toMatchSnapshot("Default Header");

  // Click the button
  const button = screen.getByRole("button", { name: "Login / Register" });
  expect(button).toBeInTheDocument();

  await userEvent.click(button);
  expect(push).toHaveBeenCalledWith("/login");
});

it("should launch logout when button is pressed", async () => {
  useZustand.setState({
    user: {
      username: "test",
      id: "test",
    },
  });

  const { push } = (useRouter as jest.Mock)();
  const { container } = render(<TaskList />);
  expect(container).toMatchSnapshot("Logged in Header");

  // Click the button
  const button = screen.getByRole("button", { name: "Logout" });
  expect(button).toBeInTheDocument();

  await userEvent.click(button);
  expect(clearCookies).toHaveBeenCalled();
  expect(push).toHaveBeenCalledWith("/login");

  useZustand.setState({});
});
