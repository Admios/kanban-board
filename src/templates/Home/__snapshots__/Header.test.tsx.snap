// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should launch login when button is pressed: Default Header 1`] = `
<div>
  <div
    class="css-1kxrbkv"
  >
    <header
      class="css-1utx3w7"
    >
      <button
        class="chakra-button css-4xx2wk"
        type="button"
      >
        Login / Register
      </button>
      <button
        class="chakra-button css-4xx2wk"
        type="button"
      >
        Add Column
      </button>
      <button
        class="chakra-button css-4xx2wk"
        type="button"
      >
        Create 10 random tasks
      </button>
    </header>
    <h2
      class="chakra-heading css-tiraa3"
    >
      Board
    </h2>
    <div
      class="css-erspa2"
    />
  </div>
</div>
`;

exports[`should launch logout when button is pressed: Logged in Header 1`] = `
<div>
  <div
    class="css-1kxrbkv"
  >
    <header
      class="css-1utx3w7"
    >
      <span
        class="css-1qm7lvz"
      >
        Your Username is: 
        test
      </span>
      <button
        class="chakra-button css-4xx2wk"
        type="button"
      >
        Logout
      </button>
      <button
        class="chakra-button css-4xx2wk"
        type="button"
      >
        Add Column
      </button>
      <button
        class="chakra-button css-4xx2wk"
        type="button"
      >
        Create 10 random tasks
      </button>
    </header>
    <h2
      class="chakra-heading css-tiraa3"
    >
      Board
    </h2>
    <div
      class="css-erspa2"
    />
  </div>
</div>
`;
